Refactor: Move `UpdateCoordinatorStateAndPrepareCluster` duty out of `DruidCoordinator`

kfaraz opened the pull request:
### Motivation
- Clean up `DruidCoordinator` and move methods to classes where they are most relevant

### Changes:
- No functional change
- Add class `PrepareBalancerAndLoadQueues` which performs the same duty as `UpdateCoordinatorStateAndPrepareCluster`
- Move creation, stopping and removal of `LoadQueuePeon`s from `DruidCoordinator` to `LoadQueueTaskMaster`
- Make `BalancerStrategyFactory` an abstract class and move handling of the balancer executor here
- Move reporting of used segment stats and historical capacity stats from `CollectSegmentAndServerStats` to `PrepareBalancerAndLoadQueues` as that is where the `DruidCluster` is created.
- Move reporting of unavailable and under-replicated segment stats from `CollectSegmentAndServerStats` to `UpdateReplicationStatus` duty


---

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

findingrish commented on server/src/main/java/org/apache/druid/server/coordinator/duty/PrepareBalancerAndLoadQueues.java:
```diff
@@ -0,0 +1,198 @@
+ *   <li>Initializes the {@link BalancerStrategy} for the run.</li>
+ * </ul>
+ */
+public class PrepareBalancerAndLoadQueues implements CoordinatorDuty
```
Is `UpdateCoordinatorStateAndPrepareCluster` deprecated now?

kfaraz replied:
The old class has been removed completely and the same content has been moved here.

findingrish replied:
Oh I see, it is nested in `DruidCoordinator` class.

findingrish commented on server/src/main/java/org/apache/druid/server/coordinator/loading/LoadQueueTaskMaster.java:
```diff
@@ -81,6 +91,50 @@ public LoadQueuePeon giveMePeon(ImmutableDruidServer server)
+
+  public void stopPeonsForDisappearedServers(List<ImmutableDruidServer> servers)
+  {
+    final Set<String> disappearedServers = Sets.newHashSet(loadManagementPeons.keySet());
```
From the method name I assumed that we would stopping peons for the sets of servers passed, instead it seems to be doing the opposite. Can you please add a comment?

kfaraz replied:
Yeah, fair point. I didn't like the structure of this method either.
Please let me know if there are any further comments. I will address all of them together.

AmatyaAvadhanula commented on server/src/main/java/org/apache/druid/server/coordinator/loading/LoadQueueTaskMaster.java:
```diff
@@ -81,6 +91,50 @@ public LoadQueuePeon giveMePeon(ImmutableDruidServer server)
+      LoadQueuePeon peon = loadManagementPeons.remove(server);
+      peon.stop();
+    }
+    loadManagementPeons.clear();
```
Nit: This may not be needed.

AmatyaAvadhanula commented on server/src/main/java/org/apache/druid/server/coordinator/loading/LoadQueueTaskMaster.java:
```diff
@@ -45,6 +53,8 @@ public class LoadQueueTaskMaster
   private final ZkPathsConfig zkPaths;
   private final boolean httpLoading;
 
+  private final ConcurrentHashMap<String, LoadQueuePeon> loadManagementPeons = new ConcurrentHashMap<>();
```
If stopAndRemoveAllPeons is called by the Coordinator when it is no longer the leader and the `PrepareBalancerAndLoadQueues` duty which was scheduled previously is running simultaneously, could there be problems?
(I think not since the coordinator would not run any further duties)

kfaraz commented:
Thanks for the review, @AmatyaAvadhanula , @findingrish . I have addressed the feedback:
- Guarded `loadMangementPeons` in `LoadQueueTaskMaster` so that once leadership is relinquished, new peons are not created
- Merge methods `stopPeonsForDisappearedServers` and `startPeonsForNewServers` into  `resetPeons`.

AmatyaAvadhanula reviewed (approved):
Thanks @kfaraz, LGTM! +1 after all the checks pass
